src/AWSLambda/Events/APIGateway.hs

```diff
@@ -0,0 +1,182 @@
+requestBodyEmbedded = requestBody . mapping unEmbed
+
+-- | Get the binary (decoded Base64) request body, if there is one
+requestBodyBinary :: Getter (APIGatewayProxyRequest Base64) (Maybe ByteString)
+requestBodyBinary = requestBody . mapping _Base64
+
+data APIGatewayProxyResponse body = APIGatewayProxyResponse
+  { _agprsStatusCode :: !Int
+  , _agprsHeaders    :: !(HashMap HeaderName HeaderValue)
+  , _agprsBody       :: !(Maybe (TextValue body))
+  } deriving (Eq, Show, Generic)
+
+instance ToText body => ToJSON (APIGatewayProxyResponse body) where
+  toJSON = genericToJSON $ aesonDrop 6 camelCase
+
+instance FromText body => FromJSON (APIGatewayProxyResponse body) where
+  parseJSON = genericParseJSON $ aesonDrop 6 camelCase
+
+$(makeLenses ''APIGatewayProxyResponse)
+
+response :: Int -> APIGatewayProxyResponse body
+response statusCode = APIGatewayProxyResponse statusCode HashMap.empty Nothing
+
+responseOK :: APIGatewayProxyResponse body
+responseOK = response 200
+
+responseNotFound :: APIGatewayProxyResponse body
+responseNotFound = response 404
+
+responseBadRequest :: APIGatewayProxyResponse body
+responseBadRequest = response 400
+
+responseBody :: Setter' (APIGatewayProxyResponse body) (Maybe body)
+responseBody = agprsBody . at () . mapping unTextValue
+
+responseBodyEmbedded :: Setter' (APIGatewayProxyResponse (Embedded body)) (Maybe body)
+responseBodyEmbedded = responseBody . mapping unEmbed
+
+responseBodyBinary :: Setter' (APIGatewayProxyResponse Base64) (Maybe ByteString)
+responseBodyBinary = responseBody . mapping _Base64
+
+-- | Process incoming events from @serverless-haskell@ using a provided
+-- function.
+--
+-- This is a specialisation of lambdaMain for API Gateway.
+--
+-- The handler receives the input event given to the AWS Lambda function, and
+-- its return value is returned from the function.
+--
+-- This is intended to be used as @main@, for example:
+--
+-- > import AWSLambda.Events.APIGateway
+-- > import Control.Lens
+-- > import Data.Aeson
+-- > import Data.Aeson.Embedded
+-- >
+-- > main = apiGatewayMain handler
+-- >
+-- > handler :: APIGatewayProxyRequest (Embedded Value) -> IO (APIGatewayProxyResponse (Embedded [Int]))
+-- > handler request = do
+-- >   putStrLn "This should go to logs"
+-- >   print $ request ^. requestBody
+-- >   pure $ responseOK & responseBodyEmbedded ?~ [1, 2, 3]
+--
+-- The type parameters @reqBody@ and @resBody@ represent the types of request and response body, respectively.
+-- The @FromText@ and @ToText@ contraints are required because these values come from string fields
+-- in the request and response JSON objects.
+-- To get direct access to the body string, use @Text@ as the parameter type.
+-- To treat the body as a stringified embedded JSON value, use @Embedded a@, where @a@ has the
+-- appropriate @FromJSON@ or @ToJSON@ instances.
+-- To treat the body as base 64 encoded binary use @Base64@.
```

Comment:
Perhaps add examples of defining `handler` for each of these cases, to reduce friction?